tests/dragonfly/conftest.py

```diff
@@ -334,38 +337,80 @@ def with_ca_tls_client_args(with_tls_client_args, with_tls_ca_cert_args):
     return args
 
 
-def copy_failed_logs_and_clean_tmp_folder(report):
+def copy_failed_logs_and_clean_tmp_folder():
     failed_path = "/tmp/failed"
     path_exists = os.path.exists(failed_path)
     if not path_exists:
         os.makedirs(failed_path)
 
-    if os.path.isfile("/tmp/last_test_log_files.txt"):
+    last_log_file = open("/tmp/failed_list.txt", "r")
+    files = last_log_file.readlines()
+    for file in files:
+        # copy to failed folder
+        shutil.copy(file.rstrip("\n"), failed_path)
+
+    # Clean up everything
+    last_log_file = open("/tmp/failed_list.txt", "w").close()
+    last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
+
+
+def pytest_exception_interact(node, call, report):
+    if report.failed:
+        # To print the test that currently failed
         last_log_file = open("/tmp/last_test_log_files.txt", "r")
+        # Global tracking of all failed tests/logs
+        failed_list = open("/tmp/failed_list.txt", "a")
         files = last_log_file.readlines()
         logging.error(f"Test failed {report.nodeid} with logs: ")
         for file in files:
-            # copy to failed folder
+            failed_list.write(file)
             file = file.rstrip("\n")
             logging.error(f"🪵🪵🪵🪵🪵🪵 {file} 🪵🪵🪵🪵🪵🪵")
-            shutil.copy(file, failed_path)
-
-
-def pytest_exception_interact(node, call, report):
-    if report.failed:
-        copy_failed_logs_and_clean_tmp_folder(report)
-
 
-@pytest.fixture(autouse=True)
+        # Clean it
+        last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
+        global TEST_FAILED
+        TEST_FAILED = True
+
+
+# Double consider any change to this function because the order
+# of SetUp and TearDown might cause improper copy of the logs
+# because the instance has not yet stopped. For example, df_factory
+# has session scope so it won't be called unless the session has ended.
+# On the other hand, df_local_factory has function scope, and therefore
+# it is executed strictly before this (so the instance is stoped and we can
+# safely copy the logs without loosing some of them).
+# autouse means that this is run for all tests in the module.
+# scope="session" allows to create the file only once and clean it
+# at the end respecting the evaluation order mentioned above
+@pytest.fixture(autouse=True, scope="session")
 def run_before_and_after_test():
-    # Setup: logic before any of the test starts
-    # Empty the log on each run
-    last_log_file = open("/tmp/last_test_log_files.txt", "w")
-    last_log_file.close()
+    logging.info("Session start for run_before_and_after_test")
+    # Setup: at the start of the session
+    last_log_file = open("/tmp/last_test_log_files.txt", "w").close()
```

Comment:
why do we need /tmp/last_test_log_files.txt ?
can we check if the test failed when stopping the dragonfly process and if so copy its logs to /tmp/failed using self.log_files?

Reply:
> can we check if the test failed when stopping the dragonfly process && why do we need /tmp/last_test_log_files.txt ?

We can check if the test failed when stopping the instance but we loose the information `which test failed with what logs`. Basically what you suggest, would almost work reasonably well for df instances with `function` scope (it would break if within a test we did `start -> stop -> start` because we would loose logs, since the stop could succeed and since there is no failure we end up copying nothing only for df to crash after the second start and we just lost the first log here) but would spectacularly `break` for df instances/fixtures with `session scope`. Why? Because, every time we run a test, we store the name of the logs for the instances we create in `last_test_log_files` so when a test fails, we print also the `name of the logs` for that test(e.g,: https://github.com/dragonflydb/dragonfly/actions/runs/9708171685/job/26794538269#step:6:438). The number one problem of copying the logs when we `stop` is that for `session fixtures` we simply don't know when they were created and because of that we will upload the logs at the end of the session but we won't know when the test failed which logs to associate with it (because it can be that multiple logs are uploaded from different fixtures and we need to know at the point of the test failure what are the `name of the logs`. 

I do have an idea which I want to try. I want to wrap up something first and will get back very soon on this

Reply:
yep I think now we cover all the cases and we also don't need a separate PR. I summarized in comments what is being done and why :)